---
title: "Group 1 Project"
output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
  html_notebook:
    toc: true
    toc_float: true
---

## Load the libraries and data

```{r, include=FALSE}
library(clusterProfiler)
library(org.Hs.eg.db)
library(edgeR)
library(gplots)
library(tidyverse)

counts <- read.csv("Group1.csv", row.names = 1)

```

### Set the factors

```{r}
treatment_pattern <- (rep(c("Untreated", "Treatment"), times = 9))
treatment <- factor(c(treatment_pattern))

donor_pattern <- c(rep("M24", times = 6), rep("M31", times = 6), rep("M32", times = 6))
donors <- factor(c(donor_pattern))

```

### Cbind the factors
```{r}
groups <- data.frame(sample = colnames(counts), treatment, donors)

cbind(groups,colnames(counts))

```
## Part 1 – Exploratory Analysis
### 1. Visualize your library sizes.
```{r}

boxplot(colSums(counts) ~ donors)

```
```{r}
boxplot(colSums(counts) ~ treatment)
```

```{r}
boxplot(colSums(counts) ~ groups$sample)
```

```{r}
hist(log2(rowSums(counts)))
```

### 3.	Determine how many genes had zero counts and remove these genes (rows) from the data set. What does your count distribution look like before and after this?

```{r}
MinVals <- apply(counts, 1, min)
sum(MinVals == 0) 
Exp <- counts[MinVals > 0, ]
```

```{r}
ExpLog2 <- log2(Exp)
hist(rowMeans(ExpLog2), xlab = "Counts (log2)", main = "Expressed Genes")
```
### 4.	Make a boxplot of log2 raw counts after removing all genes that had zero counts. 

Plot "raw" (our log2 transformed) sample counts

```{r}
boxplot(ExpLog2, ylab = "log2 counts", main = "Raw RNA-Seq Counts", las = 2)
```

### 5.	Make a Cluster Dendrogram of the raw data.

```{r}
RawDist <- dist(t(ExpLog2), method = "euclidean")
plot(hclust(RawDist, method = "average"))
```

 Simple normalization 

```{r}
SampleMedians <- apply(ExpLog2, 2, median) # Find the median value of each column - 2 means column
GrandMedian <- mean(SampleMedians) # Take the average of those
CorrectionFactors <- GrandMedian - SampleMedians # Calculate correction factor to apply to data
CorrectionFactors
```
 These correction factors will align the medians of all of our samples
 This changes our counts just enough to correct for variability between
 samples/groups without loosing any actual effects of changes in gene expression.

 Loop through each column (sample) and fill in medians adjusted with 
 sample correction factor
```{r}
ExpNorm <- ExpLog2

for(col in colnames(ExpNorm)){
  ExpNorm[, col] <- ExpLog2[, col] + CorrectionFactors[col]
}
```

### 6.	Do a simple normalization of your data and visualize your normalized data with a boxplot of log2 raw counts and a cluster dendrogram. Did normalization change anything?
```{r}
boxplot(ExpNorm, ylab = "log2 counts", main = "Normalized Counts", las = 2)
```

 Cluster dendrogram of normalized data
```{r}
NormDist <- dist(t(ExpNorm), method = "euclidean")
plot(hclust(NormDist, method = "average"))
```
### 7. Make a PCA plot and use color to highlight groups
```{r}
PCA <- prcomp(t(NormDist))
plot(PCA$x[ , 1], PCA$x[ , 2], pch = 16)
```

 By donor
```{r}
plot(PCA$x[ , 1], PCA$x[ , 2], pch = 16, col = donors, 
     main = "Colored by Donor")
legend("bottomright", legend = unique(donors), pch = 16, col = unique(donors))
```
 By treatment
```{r}
plot(PCA$x[ , 1], PCA$x[ , 2], pch = 16, col = treatment, 
     main = "Colored by Treatment")
legend("bottomright", legend = unique(treatment), pch = 16, col = unique(treatment))
```

## Part 2. Differential Gene Expression with edgeR
## When creating the DGEList object, include the gene annotation information from the "Annotations.csv" file as your argument in “genes =”
```{r}
annotations <- read.csv(file = "Annotations.csv",
                     header = TRUE, 
                     row.names = 1)
DGE <- DGEList(counts = counts, remove.zeros = TRUE, genes = annotations)
```

## Include both treatment and donor in your experimental design matrix.
```{r}
design <- model.matrix(~ 0 + treatment + donors, data = counts)
rownames(design) <- colnames(DGE)
design
```
 Remove low expression genes 
```{r}
keep <- filterByExpr(DGE, design, min.count = 10, min.total.count = 15)
DGE <- DGE[keep, ] # Keep genes that meet our threshold!
```
 Take a look at the library sizes of each sample
```{r}
DGE$samples
```
 Normalize for different library sizes (TMM, trimmed mean of M-values)
```{r}
DGE <- calcNormFactors(DGE) # Calculates normalization factors and adds them to our DGE object
DGE$samples
```
 Calculate overall dispersion
 estimateGLMCommonDisp calculates within-group and between-group variability. 
```{r}
DGE <- estimateGLMCommonDisp(DGE, design, verbose = TRUE)
```
 Calculate dispersion trend based on gene abundance
```{r}
DGE <- estimateGLMTrendedDisp(DGE, design) 
```
 Calculate separate dispersion for each gene
```{r}
DGE <- estimateGLMTagwiseDisp(DGE, design) 
```
 Visualize dispersion
```{r}
plotBCV(DGE)

```
 Let's check that our normalization worked by plotting the log2(CPM) -- counts per million
```{r}
CPM <- cpm(DGE, normalized.lib.sizes = TRUE, log = TRUE)
CPM <- as.data.frame(CPM)
```
 Visualize normalized data
```{r}
boxplot(CPM, las = 2, ylab = "log2 CPM", main = "Normalized Data")
```
### Contrast Matrix
```{r}
contrast.matrix <- makeContrasts(PA_vs_Unt = treatmentTreatment - treatmentUntreated,
                                 levels = design)
```

## 3.	When running the glmLRT function, be careful to use the correct name to specify the comparison of interest. For the purpose of this exercise, we are interested in the treatment effects.
Fit a gene-wise negative binomial generalized linear model
```{r}
fit <- glmFit(DGE, design, contrast = contrast.matrix)
lrt <- glmLRT(fit, contrast = contrast.matrix[,"PA_vs_Unt"]) 
topTags(lrt)
```
### 4a)	How many genes are differentially expressed (p<0.05 & log2 FC > 1) between the two genotypes? 
```{r}
de <- decideTests(lrt, adjust.method = "fdr")
summary(de)
```
Setting up detag
```{r}
# MERGE ANNOTATIONS
annot <- read.csv("Annotations.csv")
# Merge your added annotation information with your results
Results <- merge(annot, Results, by.x = "ENSEMBL", by.y = "genes")

# Save your results in a .csv file that you can open with excel
write.csv(Results, file = "log2FC_PA_vs_Untreated.csv", row.names = FALSE)

# But how many of those are actually significant?
# Save the ones with p-value < 0.05 and log2FC > 1
# You can change these values to be whatever threshold you'd like
# Using the unadj. p-value and FC can produce more reproducible results than 
# using FDR alone!
DE <- Results[Results$PValue < 0.05 & abs(Results$logFC) > 1, ]

# Save those gene names as "detags"
detags <- DE$ENSEMBL

```
### 4b)  Make an MA plot of edgeR results with DE genes (p<0.05 & log2 FC > 1) highlighted in red.

```{r}
# Make an MA plot highlighting significantly differentially expressed genes
plotSmear(lrt, de.tags = detags, cex = 0.5, main = "DE Genes with Pseudomonas Treatment")
abline(h = c(-1, 1), col = "blue")
legend("bottomright", legend = "DE genes (FC > 2, P-value < 0.05)", pch = 16, 
       col = "red", bty = "n")

```

### 4c) Generate a Volcano plot of edgeR results. 

```{r}
# Make a volcano plot
plot(-log10(Results$PValue) ~ Results$logFC, 
     xlab = "log2 fold change", ylab = "-log10 p-value", 
     main = "DE Genes with Pseudomonas Treatment", xlim = c(-10, 10))
points(-log10(Results$PValue[Results$ENSEMBL %in% detags]) ~ 
         Results$logFC[Results$ENSEMBL %in% detags], 
       col = "red", pch = 16)
abline(v = c(-1, 1), col = "blue")
legend("topleft", legend = "DE genes (FC > 2, p < 0.05)", pch = 16, 
       col = "red", bty = "n")


```


### 4d) Is there an uneven number of genes that are up- or down-regulated when comparing the two genotypes?

```{r}
de <- decideTests(lrt, adjust.method = "fdr")
summary(de)
```
```{r}
write.csv(Results, file = "log2FC_PA_vs_Untreated.csv", row.names = FALSE)
```
## Part 3 – Pathway Analysis (Bonus)




If there is time left, you may try to perform pathway analysis (using R or an online tool) to see if any pathways are enriched for genes that are differentially expressed between EVPA and EV treatment.

```{r}
#Import gene expression table - from the EdgeR module
edgeRdata <- read.csv("log2FC_PA_vs_Untreated.csv")
head(edgeRdata)
```

 identify up-regulated differentially expressed (DE) genes (FDR < 0.05) 
```{r}
SigGenes <- edgeRdata[which(edgeRdata$FDR<0.05 & edgeRdata$logFC>0),"ENSEMBL"]
length(SigGenes) 
head(SigGenes)

SigGenesdown <- edgeRdata[which(edgeRdata$FDR<0.05 & edgeRdata$logFC<0),"ENSEMBL"]
length(SigGenesdown) #1032 DE genes 
head(SigGenesdown)
```

 all genes detected
```{r}
ExpGenes<-edgeRdata$ENSEMBL
```

 Only ENSEMBL gene annotation is stored in the lrt object, 
 or lrt$genes, more precisely.
 However, GO terms are defined by 
 National Human Genome Research Institute, an american institute.
 Hence GO terms use ENTREZID to annotate the database, 
 and we need to add ENTREZID annotation into lrt object.
```{r}
str(annot)
head(annot)
lrt$genes<-annot[match(lrt$genes$genes,annot$ENSEMBL),]
head(lrt$genes)
```

```{r}
GoanaResults<-goana(lrt,
                    geneid=lrt$genes$ENTREZID,
                    species="Hs",
                    FDR = .05)
```

```{r}
str(GoanaResults)
head(GoanaResults)
dim(GoanaResults) 
```

 let's adjust the p-values with FDR method
 Benjamini & Hochberg (1995) ("BH" or its alias "fdr")
```{r}
GoanaResults$FDR.Up<-p.adjust(GoanaResults$P.Up, method = "fdr")
GoanaResults$FDR.Down<-p.adjust(GoanaResults$P.Down, method = "fdr")

head(GoanaResults)
```

 How many up-regulated biological processes?
```{r}
dim(GoanaResults[which(GoanaResults$P.Up<.05 & GoanaResults$Ont=="BP"),])   #2688
dim(GoanaResults[which(GoanaResults$FDR.Up<.05 & GoanaResults$Ont=="BP"),]) #1515
```
 How many up-regulated and enriched GO terms
```{r}
dim(GoanaResults[which(GoanaResults$FDR.Up<.05),])
```

 Unrapping GoanaResult 
 Top enriched GO terms in up-regulated genes 
 There is a topGO() function to identify top enriched GO Terms.
 However, topGO() is not flexible enough to identify GO term of interest
 let's just use the great old friend - [ ] - to select what we want

 The top20 GO terms that are enriched in the up-regulated genes.
 i.e. The top 20 up-regulated GO terms
```{r}
GoanaResults[order(GoanaResults$FDR.Up)[1:20],] 
View(GoanaResults[order(GoanaResults$FDR.Up)[1:20],] )
```
The top 20 GO terms are too general terms (n>4,000)
 How many genes in each GO terms
```{r}
hist(GoanaResults$N)
hist(GoanaResults$N[GoanaResults$N<1000]) # narrow down to show the histogram with N<1000
```

 let's select those terms with fewer than 200 genes involved.
 this can help us identify GO terms with more specific annotation 

```{r}
GoanaResultsBP.SU <- GoanaResults[GoanaResults$Ont == "BP" & GoanaResults$N < 200 & 
                                    GoanaResults$FDR.Up < .05, ]

GoanaResultsBP.SU[order(GoanaResultsBP.SU$FDR.Up)[1:10], ] #now, it makes more sense
```